File: experiments/updater/go/guard.go
package main

import (
	"math"
	"regexp"
	"strconv"
	"strings"
)

// Spec: experiments/updater/shared/README.md

var leadingNumeric = regexp.MustCompile(`^([0-9]+(?:\.[0-9]+)*)`)

// TagBaseVersion strips one leading "v" and any non-numeric suffix ("v4.1-DEMO2" -> "4.1").
func TagBaseVersion(tag string) string {
	stripped := strings.TrimPrefix(tag, "v")
	if match := leadingNumeric.FindStringSubmatch(stripped); match != nil {
		return match[1]
	}
	return stripped
}

// numericParts returns the parsed parts, or ok=false if any part is not a 32-bit decimal integer.
func numericParts(version string) (parts []int, ok bool) {
	if version == "" {
		return nil, false
	}
	for _, field := range strings.Split(version, ".") {
		if field == "" || strings.Trim(field, "0123456789") != "" {
			return nil, false
		}
		value, err := strconv.ParseInt(field, 10, 64)
		if err != nil || value > math.MaxInt32 {
			return nil, false
		}
		parts = append(parts, int(value))
	}
	return parts, true
}

// CompareDottedVersions orders numeric versions before non-numeric refs.
func CompareDottedVersions(a, b string) int {
	aParts, aNumeric := numericParts(a)
	bParts, bNumeric := numericParts(b)
	switch {
	case aNumeric != bNumeric:
		if aNumeric {
			return -1
		}
		return 1
	case !aNumeric:
		return strings.Compare(a, b)
	}
	for i := 0; i < max(len(aParts), len(bParts)); i++ {
		av, bv := partAt(aParts, i), partAt(bParts, i)
		if av != bv {
			if av < bv {
				return -1
			}
			return 1
		}
	}
	return 0
}

func partAt(parts []int, i int) int {
	if i < len(parts) {
		return parts[i]
	}
	return 0
}

func VersionLessThan(a, b string) bool {
	return a != b && CompareDottedVersions(a, b) < 0
}

// Allows reports whether updating from current to selectedRef may proceed.
func Allows(current, selectedRef string, allowDowngrade bool) bool {
	if selectedRef == "edge" || allowDowngrade {
		return true
	}
	return !VersionLessThan(TagBaseVersion(selectedRef), current)
}

File: experiments/updater/csharp/Guard.cs
using System.Text.RegularExpressions;

namespace DowngradeGuard;

// Spec: experiments/updater/shared/README.md
public static partial class Guard
{
    [GeneratedRegex(@"^([0-9]+(?:\.[0-9]+)*)")]
    private static partial Regex LeadingNumeric();

    /// Strips one leading "v" and any non-numeric suffix ("v4.1-DEMO2" -> "4.1").
    public static string TagBaseVersion(string tag)
    {
        var stripped = tag.StartsWith('v') ? tag[1..] : tag;
        var match = LeadingNumeric().Match(stripped);
        return match.Success ? match.Groups[1].Value : stripped;
    }

    /// Returns the parsed parts, or null if any part is not a 32-bit decimal integer.
    private static int[]? NumericParts(string version)
    {
        if (version.Length == 0)
        {
            return null;
        }

        var fields = version.Split('.');
        var parts = new int[fields.Length];
        for (var i = 0; i < fields.Length; i++)
        {
            // int.TryParse alone would accept "+1", " 1" and "-1"; the spec only allows digits.
            if (fields[i].Length == 0 || !fields[i].All(char.IsAsciiDigit) || !int.TryParse(fields[i], out parts[i]))
            {
                return null;
            }
        }
        return parts;
    }

    /// Orders numeric versions before non-numeric refs.
    public static int CompareDottedVersions(string a, string b)
    {
        var aParts = NumericParts(a);
        var bParts = NumericParts(b);

        if ((aParts is null) != (bParts is null))
        {
            return aParts is not null ? -1 : 1;
        }
        if (aParts is null || bParts is null)
        {
            return Math.Sign(string.CompareOrdinal(a, b));
        }

        for (var i = 0; i < Math.Max(aParts.Length, bParts.Length); i++)
        {
            var av = i < aParts.Length ? aParts[i] : 0;
            var bv = i < bParts.Length ? bParts[i] : 0;
            if (av != bv)
            {
                return av.CompareTo(bv);
            }
        }
        return 0;
    }

    public static bool VersionLessThan(string a, string b) =>
        a != b && CompareDottedVersions(a, b) < 0;

    /// True if updating from currentVersion to selectedRef may proceed.
    public static bool Allows(string currentVersion, string selectedRef, bool allowDowngrade) =>
        selectedRef == "edge" || allowDowngrade ||
        !VersionLessThan(TagBaseVersion(selectedRef), currentVersion);
}
